marconi-sidechain/doc/adr/01-sdd-computation.adoc

```diff
@@ -0,0 +1,214 @@
+
+The benefits:
+
+* removal of the additional 10-16GB of memory needed to compute the `LedgerState` in `marconi-sidechain`
+* enables any chain-indexers (`cardano-db-sync`, `Kupo`, etc) to read the stored `LedgerEvent` in order to index various information like SDD and rewards without using a large amount of memory
+
+The drawbacks:
+
+* requires a change to `cardano-node` in order to store those LedgerEvents. This implies lots of discussions with the `cardano-node` team to accept such a change.
+* needs the SPO to resync the relay node from scratch in order to save the `LedgerEvent` on disk. However, such a resync will only be required once, even after an upgrade on `marconi-sidechain`.
+* need to maintain a separate fork of the `cardano-node` in the event of the node team not agreeing to merge those changes
+
+=== 3. Index SDD from the state query client of the node-to-client protocol
+
+This alternative solution takes advantage of the only way currently to get the SDD from a `cardano-node` by using the state query client of the node-to-client protocol.
+However, the query will only return the latest SDD given the slot from which the Cardano node is synced to.
+In order to take advantage of this functionality, we would need to change the deployment method of `marconi-sidechain` to something as follows:
+
+. The SPO needs to stop the relay node and delete the `cardano-node` database
+. Start `marconi-sidechain` which will wait for the node to start
+. Start the relay node
+. Once Marconi has noticed that the node has started syncing, it will start querying the local query state for the SDD for each new epoch and index it in the database.
+
+Is is important to note that any changes in the indexers of `marconi-sidechain` would require a complete resync of `marconi-sidechain`.
+As a direct consequence of this solution, the user would be required to delete the node database in order to reindex everything.
+
+The container diagram is the same as the link:#compute-sdd-by-tracking-the-ledgerstate[Compute SDD by tracking the LedgerState] solution.
+
+The benefits:
+
+* no change to `cardano-node` are required
+* uses the existing infrastructure to get the required information
+
+The drawbacks:
+
+* needs the SPO to resync the relay node from scratch
+* any change in the indexing logic of `marconi-sidechain` would require resyncing from scratch the local node (thus deleting the node database)
+* overhead of needing to continuously poll the relay node to identify when a new epoch occurs
+* race condition between node and `marconi-sidechain`. If a node syncs too fast and `marconi-sidechain` is stalled because of some heavy processing, we might miss an epoch (unlikely, but possible).
```

Comment:
@eyeinsky I've updated this following your comment. Are your issues with Solution 3 sufficiently described here?

Reply:
> I guess it would be hard to get the node (or the ledger) team to write ledger events to disk because it would be almost as if they did our work for us

Not exactly, because the data would not be queryable. Chain-indexers like Kupo or Marconi could take this data and index it in a db for fast querying.

> And if they did start to persist the events to disk, then soon enough we would show up with more requests, etc. They themselves don't care about the historic data to run the node, so for them the effort would all be for external bodies.

That's true. But there's nothing wrong with asking more to the node team. We are the customers and they are the product providers. How realistic it would be to actually get such a change accepted? I'm not sure myself. At the very least, we _could_ fork the node to provide such a functionality: which kind of goes into the idea of building custom nodes of Arnaud. That's why I think this solution could be as good as the solution for making Marconi as node. 

Don't forget, there is a major drawback of the solution of Marconi as a library: mainly that any change to the indexers requires a complete resync of the blocks and LedgerState.